### PR TITLE
♻️✨ maintenance new settings (final round)

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

This is the last of a series of PRs that progressively introduced the new-settings (i.e. moving from ``trafaret``-based config to ``settings_library``-based settings ) in the entire repo . Here is a summary of the PRs preceding this one:

1. PR #2736 unifies and extends ``settings_library`` with new common settings and utils
2. PR #2739 focuses on ``web-server``: modules with dependencies to ``trafaret`` are isolated (e.g. splitting each plugin's configurations in ``_schema.py``, ``_config.py`` ).
3. PR #2750 fixes design issue and refactors ``settings_library.base.BaseCustomSettings`` 
4. PR #2744  updates all plugin's settings in the ``web-server`` to use the new ``settings_library``. The trafaret-based config is marked as deprecated but is still used to run the app. We also introduced many runtime checks to ensure no configuration is lost with the new settings. 

Finally, this PR **completely removes the deprecated trafaret-based config and activates the new settings to run the app**  :tada: (long live to trafaret! it served us very well!). These changes also facilitate **managing the plugin mechanism** of the app using env vars. For that reason, all the plugins in the webserver were also carefully reviewed and reorganized ensuring at every step that the changes were covered by the tests. 

In order to simplify the review, some descriptions of the changes and the ideas behind follow.

### Changes in the ``webserver``

There are a lot of small changes throughout the webserver modules *but* they all follow certain guidelines. Here are the hightlights:
 
- 🗑️ ``requirements``:  ``trafaret`` removed
- 🗑️ ``config`` all ``yaml`` configurations removed
- 🗑️  all ``*_schema`` and ``*config`` modules including all config dependencies (separated during PR #2739) removed. Replaced by ``settings``
- ♻️ plugin layout unified throughout the entire service ( see skeleton in #2828 ) and actively **removed module dependencies**. A plugin consists of a group of python modules that include
   - ✨ a *settings* module that defines the plugin settings. An instance of these settings is created from env vars at startup and stored in the app settings under ``ApplicationSettings.WEBSERVER_MYPLUGIN``.  Note that the plugin can be disabled by simply settings the env var ``WEBSERVER_MYPLUGIN=null``.
   - ✨ plugin module that include *a setup function* decorated with ``@app_module_setup`` where it inits the plugin and "merges with the ``app``". Typical operations are 
      - add ``app.routes`` or ``app.middlewares``
      - appends events to ``app`` signals (``app.on_startup``, ``app.cleanup_ctx``, ...)
      - inits and stores instances in app context (``app[MY_INSTANCE] = obj``). E.g. creates clients and waits a particular service to connect.
- ♻️ Some plugins/settings were split:
  -  ``WEBSERVER_RESOURCE_MANAGER`` now does not include settings of ``WEBSERVER_GARBAGE_COLLECTOR``
  -  ``WEBSERVER_SCICRUNCH`` it's a plugin on its own
- ♻️ Some plugins/settings were merged
   - ``studies_access`` modules -> ``studies_dispatcher`` (``WEBSERVER_STUDIES_DISPATCHER``)
- ♻️  ``webserver/tests`` were carefully refactored side-by-side with the plugins. These are some of the criteria used 
    - tests of a plugin must have its name as prefix, e.g. for ``myplugin``, all associated tests must be ``test_myplugin*.py`` regardless of the folder where it is.
    - Most of our test fixtures create a ``TestClient`` (typically denoted ``client``) and connects to a ``client.app`` that is  configured in each case overriding ``app_config`` fixture that would represent the trafaret-based config file.  With the new settings design, the app *is not configured anymore from file* (i.e. ``app_config``) but it uses instead env vars, as recommended in the [12-factor app](https://12factor.net/config). In order to overcome this change, we created some tools that would transform ``trafaret``-based config into env-vars. The latter are then mocked using ``monkeypatch.setenv`` to produce the same effect on the new app setting (see ``application_settings_utils.py`` and also ``monkeypatch_setenv_from_app_config`` fixture). This approach, not only saved us from a huge refactoring but also guarantees that the ``TestClient.app`` fixtures remain the same as they were before.
    - Common/useful helpers moved to ``pytest_simcore`` (reusability and reduce local module helpers)
    - All non-code files moved to ``test/data`` (easily accessible using ``test_data_dir`` fixture)

### Changes outside the ``webserver``
Screen Shot 2022-02-19 at 20 06 40
- ``models_library``
    - 🗑️ deprecated modules ``models_library.settings.rabbit`` and ``models_library.settings.redis`` removed (replaced by new ``settings_library`` counterparts
- ``pytest_simcore``:
   -  ``helpers``:
       -   ♻️``utils_dict`` minor
       -   🐛 ``utils_docker`` fixes on ``safe_docker_infos`` due to long names produces when creating ``test_failures`` reports
       - ♻️``utils_login``: adapted to new settings for login plugin. Removed global config. Further refactoring might be necessary to access users with ``aiopg``. This is a very old fixture
       - ♻️``utils_projects`` moved some of the helpers with projects here to reduce redundancy in project test suites
   -  ♻️``rabbit_service`` fixtures to use new ``service_settings.rabbit``
   - ♻️``redis_service`` fixtures to use new ``service_settings.redis``
- ``servicelib``
   - ✨ ``application_setup`` enhances plugin initialization. Now decorator relies on application settings (instead of configs) and ensures a that **the plugin setup is only executed once**. NOTE: @GitHK @sanderegg  I will add the dependency feature we discussed in a separate PR.
- ``settings_library``
    - ✨``utils_service``: simplified composing URLs from host, port ... (see ``test_utils_service.py``)
    - ✨``utils_session``: New mixin to check fernet keys. Used in both the ``web-server`` and ``api-server`` session settings

Screen Shot 2022-02-19 at 20 33 29
- ``api-server``
   - ✨ ``settings`` uses new mixin from ``servicelib.utils_session``. 
   - ♻️ ``webserver`` module adapted to use new settings . Fixes communication due to changes in webserver
- ``director-v2``: 
     - 🗑️``/tests``  refactored due to deprecation of ``models_library.settings.rabbit`` and ``models_library.settings.redis``
- ``director``
    - 📝 ``/tests`` highlighted in doc that changes in ``pytest_simcore`` that are not python 3.6 compliant will break these tests (it [happened during this PR](https://github.com/ITISFoundation/osparc-simcore/runs/5252521258?check_suite_focus=true) and errors were quite cryptic)
- ``dynamic-sidecar``
    - 🗑️ ``/tests``  same as above

## Related issue/s

- #2008 
- ITISFoundation/osparc-issues#428

## How to test

Take any plugin (e.g. ``exporter``), then type
```cmd
pytest --pdb --ff -vv tests/**/test_exporter*.py
```
NOTE: we could create a make recipe to test plugins

## Checklist

- [x] Replace config by settings in plugins:
  - [x] ``activity/module_setup.py``  | `` pytest --pdb tests/unit/**/test_activ*.py``
  - [x] ``catalog.py`` | ``pytest --pdb -vv --ff tests/unit/**/test_catalog*.py``
  - [x] ``clusters/module_setup.py`` | ``pytest --pdb -vv tests/unit/with_dbs/**/test_clust*.py``
  - [x] ``computation.py`` 
  - [x] db.py:97
  - [x] diagnostics.py:25
  - [x] director_v2.py:19
  - [x] director/module_setup.py:22
  - [x] email.py:24
  - [x] exporter/module_setup.py:21
  - [x] garbage_collector.py:12
  - [x] groups.py:21 
  - [x] scicrunch
  - [x] login/module_setup.py:69
  - [x] meta_modeling.py:23
  - [x] products.py:102
  - [x] projects/module_setup.py:58
  - [x] publications.py:20  -> has NO TESTS!
  - [x] redis.py:102
  - [x] remote_debug.py:16
  - [x] resource_manager/module_setup.py:21
  - [x] rest.py:29
  - [x] security.py:25
  - [x] session.py:40
  - [x] socketio/module_setup.py:17 -> NO TESTS
  - [x] statics.py:186
  - [x] storage.py:17
  - [x] studies_access.py:283
  - [x] studies_dispatcher/module_setup.py:18
  - [x] tags.py:20
  - [x] tracing.py:24
  - [x] users.py:20
  - [x] version_control.py:21
- [x] version_control.py:21
- [x] rm config-based boot
- [x] rm schemas and config modules
- [x] rm config files
- [x] rm config tests / replace config with comprehensive environs
- [x] rm trafaret dependency
- [x] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: packages/pytest-simcore/src/pytest_simcore/docker_swarm.py [hunk 1]
```diff
@@ -217,13 +217,30 @@ def docker_stack(
...
+            )
+        except subprocess.CalledProcessError as err:
+            print(
+                "docker_stack failed",
+                f"{' '.join(err.cmd)}",
+                f"returncode={err.returncode}",
+                f"stdout={err.stdout}",
+                f"stderr={err.stderr}",
+            )
+            raise
```
Comment (sanderegg) [inline packages/pytest-simcore/src/pytest_simcore/docker_swarm.py:242]: nice!

File: packages/pytest-simcore/src/pytest_simcore/rabbit_service.py [hunk 2]
```diff
@@ -39,37 +39,51 @@ async def wait_till_rabbit_responsive(url: str) -> None:
...
-        host="127.0.0.1",
-        port=get_service_published_port("rabbit", testing_environ_vars["RABBIT_PORT"]),
-        channels=json.loads(testing_environ_vars["RABBIT_CHANNELS"]),
+
+    port = get_service_published_port("rabbit", testing_environ_vars["RABBIT_PORT"])
+
+    settings = RabbitSettings(
+        RABBIT_USER=testing_environ_vars["RABBIT_USER"],
+        RABBIT_PASSWORD=testing_environ_vars["RABBIT_PASSWORD"],
+        RABBIT_HOST="127.0.0.1",
```
Comment (sanderegg) [inline packages/pytest-simcore/src/pytest_simcore/rabbit_service.py:55]: should we not get the host ip here instead of 127.0.0.1? in case this is used from a container as well?

File: packages/pytest-simcore/src/pytest_simcore/rabbit_service.py [hunk 3]
```diff
@@ -39,37 +39,51 @@ async def wait_till_rabbit_responsive(url: str) -> None:
...
+
+    NOTE: Use this fixture to setup client app
+    """
+    monkeypatch.setenv("RABBIT_HOST", rabbit_settings.RABBIT_HOST)
+    monkeypatch.setenv("RABBIT_PORT", str(rabbit_settings.RABBIT_PORT))
+    monkeypatch.setenv("RABBIT_USER", rabbit_settings.RABBIT_USER)
+    monkeypatch.setenv(
+        "RABBIT_PASSWORD", rabbit_settings.RABBIT_PASSWORD.get_secret_value()
+    )
+    monkeypatch.setenv("RABBIT_CHANNELS", json.dumps(rabbit_settings.RABBIT_CHANNELS))
```
Comment (sanderegg) [inline packages/pytest-simcore/src/pytest_simcore/rabbit_service.py:79]: minor: are these monkeypatch necessary? since the rabbit_settings are already setting them? but the port ok.

File: packages/pytest-simcore/src/pytest_simcore/redis_service.py [hunk 4]
```diff
@@ -19,32 +19,46 @@
...
-        port=get_service_published_port(
-            "simcore_redis", testing_environ_vars["REDIS_PORT"]
-        ),
+    port = get_service_published_port(
+        "simcore_redis", testing_environ_vars["REDIS_PORT"]
     )
-    await wait_till_redis_responsive(str(config.dsn))
-    return config
+    # test runner is running on the host computer
+    settings = RedisSettings(REDIS_HOST="127.0.0.1", REDIS_PORT=int(port))
```
Comment (sanderegg) [inline packages/pytest-simcore/src/pytest_simcore/redis_service.py:36]: should we not set the host ip instead of 127.0.0.1? same reason as above.

File: packages/service-library/src/servicelib/aiohttp/application_setup.py [hunk 5]
```diff
@@ -114,11 +140,13 @@ def app_module_setup(
     module_name: str,
     category: ModuleCategory,
     *,
+    # TODO: use settings_name as module_name!!
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/aiohttp/application_setup.py:143]: is that still current?

File: packages/service-library/src/servicelib/aiohttp/application_setup.py [hunk 6]
```diff
@@ -114,11 +140,13 @@ def app_module_setup(
...
+    settings_name: Optional[str] = None,
     depends: Optional[List[str]] = None,
+    logger: logging.Logger = log,
+    # TODO: deprecate these options
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/aiohttp/application_setup.py:147]: is that still todo? or forgotten?

File: packages/service-library/src/servicelib/aiohttp/application_setup.py [hunk 7]
```diff
@@ -150,26 +178,17 @@ def setup(app: web.Application):
...
 
         if "setup" not in setup_func.__name__:
             logger.warning("Rename '%s' to contain 'setup'", setup_func.__name__)
 
         # metadata info
-        def setup_metadata() -> Dict[str, Any]:
+        def setup_metadata() -> SetupMetadataDict:
             return {
                 "module_name": module_name,
                 "dependencies": depends,
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/aiohttp/application_setup.py:194]: you probably need to add some asserts in front?

File: packages/settings-library/src/settings_library/utils_service.py [hunk 8]
```diff
@@ -2,14 +2,24 @@
...
 from .basic_types import PortInt
 
 DEFAULT_AIOHTTP_PORT: PortInt = 8080
 DEFAULT_FASTAPI_PORT: PortInt = 8000
 
 
+class URLPart(Enum):
+    EXCLUDE = auto()
+    OPTIONAL = auto()
+    REQUIRED = auto()
```
Comment (sanderegg) [inline packages/settings-library/src/settings_library/utils_service.py:20]: just wondering about this auto method.. just checked in the documentation. how does that work if we add another value? it has to be appended right? at least if the value goes somewhere?

File: services/web/server/src/simcore_service_webserver/clusters/plugin.py [hunk 9]
```diff
@@ -9,29 +9,22 @@
...
 
-log = logging.getLogger(__file__)
+log = logging.getLogger(__name__)
 
 
 @app_module_setup(
     "simcore_service_webserver.clusters",
     ModuleCategory.ADDON,
-    depends=["simcore_service_webserver.groups"],
+    settings_name="WEBSERVER_CLUSTERS",
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/clusters/plugin.py:23]: question: why did you remove the dependency to the groups?

File: services/web/server/src/simcore_service_webserver/application_settings_utils.py [hunk 10]
```diff
@@ -148,14 +157,15 @@ def convert_to_app_config(app_settings: ApplicationSettings) -> Dict[str, Any]:
...
+        "scicrunch": {"enabled": app_settings.WEBSERVER_SCICRUNCH is not None},
         "statics": {
             "enabled": app_settings.WEBSERVER_FRONTEND is not None
             and app_settings.WEBSERVER_STATICWEB is not None
         },
-        # NOTE:  app_settings.WEBSERVER_STUDIES_ACCESS_ENABLED did not apply
-        "studies_access": {"enabled": app_settings.WEBSERVER_STUDIES_ACCESS},
         # NOTE  app_settings.WEBSERVER_STUDIES_ACCESS_ENABLED did not apply
-        "studies_dispatcher": {"enabled": app_settings.WEBSERVER_STUDIES_DISPATCHER},
+        "studies_dispatcher": {
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/application_settings_utils.py:166]: is this still needed without trafaret?

File: services/web/server/src/simcore_service_webserver/groups.py [hunk 11]
```diff
@@ -12,22 +12,25 @@
...
     depends=["simcore_service_webserver.rest", "simcore_service_webserver.users"],
     logger=logger,
 )
 def setup_groups(app: web.Application):
 
+    assert app[APP_SETTINGS_KEY].WEBSERVER_GROUPS  # nosec
+
     # prepares scicrunch api
-    setup_scicrunch_submodule(app)
+    setup_scicrunch(app)
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/groups.py:33]: scicrunch in groups?

File: services/web/server/src/simcore_service_webserver/remote_debug.py [hunk 12]
```diff
@@ -25,15 +34,14 @@ def setup_remote_debugging(app: web.Application):
             #
             import ptvsd
 
-            REMOTE_DEBUGGING_PORT = 3000
             ptvsd.enable_attach(
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/remote_debug.py:37]: minore: now one uses debugpy instead of ptvsd (see [https://github.com/ITISFoundation/osparc-dask-gateway/blob/22d8663fc143ef3e76b4b397af1ec3ce4fcdd003/services/osparc-gateway-server/src/osparc_gateway_server/remote_debug.py](https://github.com/ITISFoundation/osparc-dask-gateway/blob/22d8663fc143ef3e76b4b397af1ec3ce4fcdd003/services/osparc-gateway-server/src/osparc_gateway_server/remote_debug.py))

**Review by sanderegg — APPROVED** [review]
RIP trafaret! very nice work that will simplify a lot the development on the webserver! Thanks!

**Review by mguidon — APPROVED** [review]
Outstanding refactoring work in all 5 PRs. Thanks!

File: services/web/server/src/simcore_service_webserver/remote_debug.py (hunk 12 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/remote_debug.py:37]: Good idea. I prefer doing that once for the entire repo in a separate PR. Added #2844

File: services/web/server/src/simcore_service_webserver/clusters/plugin.py (hunk 9 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/clusters/plugin.py:23]: my mistake. 
The dependency mechanism will be also improved. For the moment is just used as an "assertion" mechanism at startup

File: packages/settings-library/src/settings_library/utils_service.py (hunk 8 above)
Comment (pcrespov) [reply packages/settings-library/src/settings_library/utils_service.py:20]: it just adds a different value.
the way i use it is: if the enum is not going to be stored , then I just care that they are different and leave it to ``auto`` to choose them. If it is persisted, then I am the one settings the values explicitly

File: packages/service-library/src/servicelib/aiohttp/application_setup.py (hunk 7 above)
Comment (pcrespov) [reply packages/service-library/src/servicelib/aiohttp/application_setup.py:194]: What do you mean?

File: packages/service-library/src/servicelib/aiohttp/application_setup.py (hunk 5 above)
Comment (pcrespov) [reply packages/service-library/src/servicelib/aiohttp/application_setup.py:143]: Right, this will come in a separate PR where I will completely remove all the config. There are some extra functionality coming and I preferred doing it separately (this way I do not need to redo the tests)

File: packages/service-library/src/servicelib/aiohttp/application_setup.py (hunk 6 above)
Comment (pcrespov) [reply packages/service-library/src/servicelib/aiohttp/application_setup.py:147]: same as above. I added more doc in the TODO

File: services/web/server/src/simcore_service_webserver/application_settings_utils.py (hunk 10 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/application_settings_utils.py:166]: yes. what i did was moving studis_access into studies_dispatch. These utils here is simply to convert the configs in the tests into env-vars. It is not used in the code. There is an explanation about these utils in the description above under the tests section.

File: services/web/server/src/simcore_service_webserver/groups.py (hunk 11 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/groups.py:33]: i separated  it.

File: packages/pytest-simcore/src/pytest_simcore/rabbit_service.py (hunk 2 above)
Comment (pcrespov) [reply packages/pytest-simcore/src/pytest_simcore/rabbit_service.py]: good point.

**Review by GitHK — APPROVED** [review]
Thanks for now. Will have a better look at it later in the future.

File: packages/pytest-simcore/src/pytest_simcore/rabbit_service.py (hunk 3 above)
Comment (pcrespov) [reply packages/pytest-simcore/src/pytest_simcore/rabbit_service.py:79]: This was the design in place. The way I understood is:
- start stack and wait until is up
- wait until given service is responsive (e.g. redis, rabbit, etc)
- fill settings to create a client for that service
and then, there are two types of fixtures
1. ``*_service`` guarantees services is up and running, and env vars to use it are in monkeypatched
2. ``*_client``  guarantees services is up and running, and returns a client ready to interact with it

File: packages/pytest-simcore/src/pytest_simcore/redis_service.py (hunk 4 above)
Comment (pcrespov) [reply packages/pytest-simcore/src/pytest_simcore/redis_service.py]: done

**Comment by mrnicegyu11** [comment]
Note: This PR makes setting the Env-Var `REST_SWAGGER_API_DOC_ENABLED` necessary if the swaggerAPI docs are to be exposed

**Comment by pcrespov** [comment]
> Note: This PR makes setting the Env-Var `REST_SWAGGER_API_DOC_ENABLED` necessary if the swaggerAPI docs are to be exposed

yes. @mrnicegyu11 check rules of the reverse proxy as well in case the ``/dev/doc`` is filtered out
